platform/pulp/download/batch.py

```diff
@@ -0,0 +1,535 @@
+        >>>
+        >>> with Batch(downloads) as batch:
+        >>>     for plan in batch():
+        >>>         if plan.succeeded:
+        >>>             # Use the downloaded file \o/
+        >>>         else:
+        >>>             # Log something
+        >>>
+        >>> # Not using context manager
+        >>>
+        >>> batch = Batch(downloads)
+        >>> try:
+        >>>     for plan in batch():
+        >>>         if plan.succeeded:
+        >>>             # Use the downloaded file \o/
+        >>>         else:
+        >>>             # Log something
+        >>> finally:
+        >>>     batch.shutdown()
+        >>>
+    """
+
+    def __init__(self, downloads, concurrent=CONCURRENT, backlog=BACKLOG, **context):
+        """
+        Args:
+            downloads (generator): A generator of downloads.
+            concurrent (int): The number of downloads to execute in concurrently.
+            backlog (int): The number of downloads kept in memory.
+            **context (dict): Download context properties to be shared with downloads
+                such as sessions and auth tokens.  See the specific download documentation
+                for details on supported properties.
+
+        Raises:
+            ValueError: concurrent less than 2 or backlog is less than concurrent.
+        """
+        super(Batch, self).__init__()
+        self.downloads = downloads
+        self.concurrent = concurrent
+        self.iterator = PlanIterator(backlog)
+        self.executor = BatchExecutor(concurrent=concurrent, backlog=backlog)
+        self.context = Context(**context)
+        self.feeder = DownloadFeeder(self)
+        self.is_shutdown = False
+        if concurrent < 2:
+            raise ValueError(_('concurrent may not be < 2'))
+        if backlog < concurrent:
+            raise ValueError(_('backlog may not be < concurrent'))
+
+    def download(self):
+        """
+        Execute all of the downloads.
+
+        Returns:
+            PlanIterator: A plan iterator.
+                The iterator will yield the download `Plan`s in the order completed.
+        """
+        log.debug(_('%(batch)s - download started'), {'batch': self})
+        self.feeder.start()
+        return self.iterator
+
+    def __call__(self):
+        """
+        Execute all of the downloads.
+
+        Returns:
+            PlanIterator: A plan iterator.
+                The iterator will yield the download `Plan`s in the order completed.
+        """
+        return self.download()
+
+    def shutdown(self):
+        """
+        End processing and shutdown the feeder and the thread pool.
+
+        Notes:
+            This must be called to prevent leaking resources unless the Batch
+            is used as a context manager.
+            >>>
+            >>> with Batch(..) as batch:
+            >>>    # ...
+        """
+        if self.is_shutdown:
+            return
+        self.is_shutdown = True
```

Comment:
The `is_shutdown` attributes indicate that `shutdown()` has been called and that _shutdown_ is, at least, in progress.  Mainly used internally within the package.  I can update the docstrings on those attributes or consider renaming them (though, not sure to what).  If the feeder or the executor fails to shutdown an exception will be raised and there really isn't a remedial action.